### ISSUE TITLE
Wrong format of CommandLine property of Get-Process cmdlet

### ISSUE THREAD

**Issue by 237dmitry** [PowerShell/PowerShell#18705]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

CommandLine property returns the non-separated string. For example:

```powershell
(Get-Process -Id $PID).CommandLine
```
return `pwsh-NoLogo`

### Expected behavior

```console
pwsh -NoLogo
```

### Actual behavior

```console
pwsh-NoLogo
```

### Error details

```console
no errors
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.3.0
PSEdition                      Core
GitCommitId                    7.3.0
OS                             Linux 6.0.10-arch2-1 #1 SMP PREEMPT_DYNAMIC Sat, 26 Nov 2022 16:51:18 +0000
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

![ArcoLinux_2022-12-02_00-20-03](https://user-images.githubusercontent.com/78153320/205161482-91103a60-2cc8-4024-867b-58b263a117b6.png)

I think this depends on dotnet:

![ArcoLinux_2022-12-02_00-30-18](https://user-images.githubusercontent.com/78153320/205163373-8eb4f435-3fe4-4ed4-bfa9-8979562f2732.png)

**Comment by jborean93** [comment]
The `CommandLine` property is added by PowerShell as an ETS property and is not something in dotnet

```powershell
$member = Get-Process -Id $pid | Get-Member -Name CommandLine
$member
$member.Definition
```

We can see the member definition of this ScriptProperty being

```powershell
System.Object CommandLine {get=
                        if ($IsWindows) {
                            (Get-CimInstance Win32_Process -Filter "ProcessId = $($this.Id)").CommandLine
                        } elseif ($IsLinux) {
                            Get-Content -LiteralPath "/proc/$($this.Id)/cmdline"
                        }
                    ;}
```

So on Linux it is simply getting the contents of `/prop/$processId/cmdline`. It looks like we need to update the logic to properly read the file as it's using a null byte to delimit each argument rather than a space. For example the pid `75688` was started as `pwsh -NoProfile -NoLogo` but we can see the raw bytes of this file being:

```powershell
PS /home/jborean> Format-Hex -Path /proc/75688/cmdline

   Label: /proc/75688/cmdline

          Offset Bytes                                           Ascii
                 00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
          ------ ----------------------------------------------- -----
0000000000000000 70 77 73 68 00 2D 4E 6F 4C 6F 67 6F 00 2D 4E 6F pwsh -…
0000000000000010 50 72 6F 66 69 6C 65 00                         Profil…
```

Each argument is separated by `00` and not a space `20`. This makes sense as arguments themselves could have a space and Linux doesn't actually have a string command line format (they take an array of args).

A naive fix is to just replace the null byte with a space (or split by null and join by space)

```powershell
(Get-Content -LiteralPath "/proc/$($this.Id)/cmdline") -replace "`0", " "
``` 

The problem with this approach is if an argument itself contains a space and trying to convert the array of arguments you now have to how it was started by the shell. Should it;

* Just not care and say this is a best guesstimate
* Try and convert it to a valid command line that can be used in pwsh (single quotes or double quotes)
* Use bash argument quoting rules, or figure out what started it and try and use those rules

**Comment by mklement0** [comment]
Great analysis, @jborean93.

This issue is a a duplicate of

*  https://github.com/PowerShell/PowerShell/issues/13944

Related:

* https://github.com/PowerShell/PowerShell/issues/13943

**Comment by 237dmitry** [comment]
> The CommandLine property is added by PowerShell as an ETS property and is not something in dotnet

Why then is the identical result in `[system.diagnostics.process]`?

I understood about `/proc/ID/cmdline`.  Strange that developers did not replace ``` `u{0} ``` with ``` `u{32} ```

**Comment by jborean93** [comment]
> Why then is the identical result in [system.diagnostics.process]?

This returns the same object as `Get-Process`, the ETS properties are automatically added on top of the `System.Diagnostics.Process` type regardless of where it is from.

> Strange that developers did not replace `u{0} with `u{32}

Because the value in that file is essentially the serialized form of the argv array when the process was created. This array is essentially an array of null terminated strings. If the question was why PowerShell didn't convert it it was mostly because they didn't realise that was the format. Still just blindly converting to a space may give a value people won't expect as it's not always going to be the string people put in their shell.

**Comment by 237dmitry** [comment]
`s/32/20`

**Comment by jborean93** [comment]
I'm curious what would would you expect it to show in the case of arguments that have a space. Say you can `/bin/test argument 'quoted value'` in Bash what would you expect back? Should it be

* `/bin/test argument quoted value`
* `/bin/test argument 'quoted value'`
* `/bin/test argument "quoted value"`
* Something else?

What about if the argument had embedded single or double quotes as the value, the escaping method is highly dependent on the shell used.

It's hard to convert the raw argv value in the `/proc/PID/cmdline` file back to what you actually ran s that information isn't preserved anywhere. So while PowerShell could do the naive approach, or quote it in a way that works for it, it's not necessarily going to be what people may expect if they ran it from a different shell.

**Comment by jborean93** [comment]
I'm not protecting anyone, and it's not a rhetorical question. Linux, unlike Windows does not store the command line of a process as a string, it stores it as an array of values. This means the string people would have typed into their shell is not something you can necessarily get back. The shell used will convert the string to an array of arguments and that's what is stored in the file. As I mentioned, if you were to do `/bin/test argument 'quoted value'` in bash the cmdline file will be `/bin/test\0argument\0quoted value\0`. Notice how the single quotes that were used in the shell are no longer present.

This brings up my question, how should PowerShell interpret the raw value of null terminated strings into a command line that people expect. How does it know that the original command line used single quotes vs double quotes? Should it care about it at all? Should it just return an array of strings instead? These are all things that would need to be answered before this is fixed and the best people to answer that is the people who are going to use this value, like yourself.

**Comment by jborean93** [comment]
Just as an FYI, `ps` takes in option one and doesn't deal with quotes. Running the following in bash `pwsh -NoExit -Command '$pid; "testing abc"'` and the doing `ps -aux | grep pwsh` in another terminal will give you:

```bash
jborean:~$ ps -aux | grep pwsh
jborean   259054  8.0  0.1 274836356 176756 pts/0 Sl+ 14:20   0:00 pwsh -NoExit -Command $pid; "testing abc"
jborean   259506  0.0  0.0 222020  2308 pts/1    S+   14:21   0:00 grep --color=auto pwsh
```

Notice how the single quotes of the original process are gone, ps is just going the simple route of separated by space and not worrying about trying to re-add the quotes back in.

**Comment by 237dmitry** [comment]
```
15796
testing: The term 'testing' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```

> how should PowerShell interpret the raw value of null terminated strings into a command line that people expect.

```
pwsh -NoExit -Command $pid; "testing abc"
```

**Comment by mklement0** [comment]
Hear, hear, @jborean93.

To boil it down to the essentials:

* On Unix, there is no such thing as a _process command line_ - there's only an _array of verbatim arguments_ that a process received, and that's all the system knows about. On Unix, a command line is purely a _shell_ construct, and recreating a command line is _speculative_ and involves an ultimately _arbitrary_ choice of a  particular shell's syntax (although `/bin/sh` would be a natural candidate).

  * As the `ps` examples show, `ps` doesn't even make such an attempt, and simply _space-concatenates the verbatim arguments_; in other words: except in simple cases, this "command line" doesn't even work in POSIX-compatible shells, and is as such primarily useful to the _human observer_, as an _approximation_ of what was originally invoked.

* On Windows, there is _only_ a _process command line_, and it is the only thing the system can report. Given that it is up to each process to break its command line down into arguments, any process-external attempt to report a an _array of verbatim arguments_ would be _speculative_ (although using what the Microsoft C/C++ parsing rules expect would be a natural candidate).

See [here](https://github.com/PowerShell/PowerShell/issues/13944#issuecomment-1334750388) for further discussion.

**Comment by 237dmitry** [comment]
> On Unix, there is no such thing as a process command line 

```
$ ps -e -o command | grep pwsh
tilix --window-style=borderless -p pwsh
pwsh -NoLogo
grep pwsh
```

**Comment by mklement0** [comment]
My previous comment explains how this  "command line" is being constructed and you can also infer it from @jborean93's examples and comment; to recap: 

It is _not_ a shell command line, it is the _space-concatenated list of the verbatim arguments_ passed to the process, _which is all the system knows about_, and as such _unrelated to any real shell command line_, which could have come from _any_ shell, and - except in the simple case of all-space-less, shell-metacharacter-free arguments - won't work as a command line in _any_ shell.

**Comment by mklement0** [comment]
To put it differently:

PowerShell _could_ follow the example of `ps` and perform the same, simple space concatenation.

This would certainly make the output more _readable_, but comes at the expense of knowing the true argument values / boundaries.

But the questions are:

* Is it better to at least emulate _some_ shell's syntax in recreating a hypothetical command line? If so, should it be PowerShell's syntax or `/bin/sh`'s syntax?

* Should the array of verbatim arguments be reported via a _different_  property?

**Comment by 237dmitry** [comment]
But result is correct. Why not do it right out of the box...

```powershell
 $ ((gps).CommandLine | sls 'pwsh' -raw) -replace "`u{0}", "`u{20}"
pwsh -NoLogo 
tilix --window-style=borderless -p pwsh 
```

**Comment by mklement0** [comment]
No, the result isn't _generally_ correct - what you're showing is in effect the simple space concatenation - see @jborean93's `pwsh -NoExit -Command '$pid; "testing abc"'` command for an example of where that breaks down.

**Comment by 237dmitry** [comment]
> No, the result isn't generally correct

And what? It's not even funny anymore... Either floating-point numbers in the `pscustombject` are rounded to thousandths, or Microsoft cannot correctly deal with processes, or users cannot find the necessary libraries, tomorrow we will learn new interesting news.

Here is the quintessence:
```powershell
 $ [pscustomobject] @{ a = 1.0; b = 1.12 }

    a     b
    -     -
1,000 1,120
```

**Comment by jborean93** [comment]
I think we just need to take a step back. There’s obviously a problem as you’ve pointed out. The issue is how do we fix it. There are many options available to us but we need input from the pwsh team on which route to take. Anything more will just result in frustration between us all and we certainly don’t want that.

**Comment by 237dmitry** [comment]
> I think we just need to take a step back.

Ok. I closed issue. Because this is not productive.

**Comment by mklement0** [comment]
Indeed, @237dmitry, your contributions were not productive - thank you for exiting the discussion.

For those willing to contribute productively, the discussion continues [here](https://github.com/PowerShell/PowerShell/issues/13944#issuecomment-1334773344).

**Comment by 237dmitry** [comment]
To solve the problem wrote the _**Pseudo Proxy Function**_. I could not change or set new `ScriptProperty` in _**Real  Proxy Function**_

<details>
  <summary>Get-Process (Added ScriptProperty "Line")</summary>

```powershell
function Get-Process
{
    [CmdletBinding(DefaultParameterSetName='Name', HelpUri='https://go.microsoft.com/fwlink/?LinkID=2096814', RemotingCapability='SupportedByCommand')]
    param (

        [Parameter(ParameterSetName='Name', Position=0, ValueFromPipelineByPropertyName=$true)]
        [Parameter(ParameterSetName='NameWithUserName', Position=0, ValueFromPipelineByPropertyName=$true)]
        [Alias('ProcessName')]
        [ValidateNotNullOrEmpty()]
        [string[]]
        ${Name},

        [Parameter(ParameterSetName='Id', Mandatory=$true, ValueFromPipelineByPropertyName=$true)]
        [Parameter(ParameterSetName='IdWithUserName', Mandatory=$true, ValueFromPipelineByPropertyName=$true)]
        [Alias('PID')]
        [int[]]
        ${Id},

        [Parameter(ParameterSetName='InputObject', Mandatory=$true, ValueFromPipeline=$true)]
        [Parameter(ParameterSetName='InputObjectWithUserName', Mandatory=$true, ValueFromPipeline=$true)]
        [System.Diagnostics.Process[]]
        ${InputObject},

        [Parameter(ParameterSetName='NameWithUserName', Mandatory=$true)]
        [Parameter(ParameterSetName='IdWithUserName', Mandatory=$true)]
        [Parameter(ParameterSetName='InputObjectWithUserName', Mandatory=$true)]
        [switch]
        ${IncludeUserName},

        [Parameter(ParameterSetName='Name')]
        [Parameter(ParameterSetName='Id')]
        [Parameter(ParameterSetName='InputObject')]
        [ValidateNotNull()]
        [switch]
        ${Module},

        [Parameter(ParameterSetName='Name')]
        [Parameter(ParameterSetName='Id')]
        [Parameter(ParameterSetName='InputObject')]
        [Alias('FV','FVI')]
        [ValidateNotNull()]
        [switch]
        ${FileVersionInfo}
    )

    $wrappedCmd = $ExecutionContext.InvokeCommand.GetCommand('Microsoft.PowerShell.Management\Get-Process', [System.Management.Automation.CommandTypes]::Cmdlet)

    $ps = & $wrappedCmd @PSBoundParameters

    $processes =
    foreach ($p in $ps)
    {
        $p = Add-Member -InputObject $p `
                        -MemberType ScriptProperty `
                        -Name Line `
                        -Value { $this.CommandLine -replace "`u{0}","`u{20}" } `
                        -PassThru
        $p
    }

    return $processes
}
```

</details>

**Comment by mklement0** [comment]
Note that the simple space concatenation of the verbatim arguments  that you were looking for (as also reported by `ps`) now HAS been implemented, as at least _some_ improvement over the current behavior (possibly to be improved later): 

 * https://github.com/PowerShell/PowerShell/pull/18710

I presume this will be available in v7.3.1 (as well as its _preview_ versions).

**Comment by 237dmitry** [comment]
I apologize for the nervous behavior.
